### DISCUSSION TITLE
[C++] Unclear documentation for adding schemas

### DISCUSSION THREAD

**Discussion by jarkenau** [foxglove/mcap#1102] [General] [answered]
The documentation for C++ includes a [minimal example](https://github.com/foxglove/mcap/tree/main/cpp) on how to add schemas. For writing a simple `std_msgs/String` it can be done like this:

```C++
...
// Register a Schema
mcap::Schema stdMsgsString("std_msgs/String", "ros1msg", "string data");
writer.addSchema(stdMsgsString);

// Register a Channel
mcap::Channel chatterPublisher("/chatter", "ros1", stdMsgsString.id);
writer.addChannel(chatterPublisher);
...
```
Where `string data` describes the layout of the message. How should the string look like for more complex & nested messages, e.g `sensor_msgs/Image`?

**Comment by foxhubber** [comment]
Linear: [FG-1855](https://linear.app/foxglove/issue/FG-1855/[c]-unclear-documentation-for-adding-schemas-[foxglovemcap-810])

**Comment by jtbandes** [comment, marked as answer]
The format for ROS schemas is documented here: https://mcap.dev/specification/appendix.html

**Comment by jarkenau** [comment]
Thanks for the hint :+1: . But it still won't work as expected. As described in the docs [here](https://mcap.dev/specification/appendix.html#ros1msg-data-format) I used `gendeps --cat` to generate the description and saving `sensor_msgs/Image` now works just fine. If I open the written `.mcap` file in foxglove studio, the image panel throws the following error `Cannot read properties of undefined (reading 'stamp')`, am I missing something?

**Reply by jtbandes** [reply]
Can you please share the mcap file you created so we can take a look? If you compress into a .zip or a .tar.gz then you can attach it here.

**Reply by jarkenau** [reply]
The images are randomly generated by a node (just for testing purposes). Here is the resulting MCAP file:
[mcap-file.zip](https://github.com/foxglove/community/files/10725086/mcap-file.zip)

**Reply by jtbandes** [reply]
Looking at your schema data, it appears the lines are concatenated without any `\n` newlines, so the schema is treated as one large comment, and thus the app doesn't think it contains any fields.

<img width="1334" alt="image" src="https://user-images.githubusercontent.com/14237/218543556-527088bf-499d-4ede-9335-8eb789036618.png">

<img width="924" alt="image" src="https://user-images.githubusercontent.com/14237/218543984-57ef44d4-8286-4f97-88f9-5d135e6d1e32.png">

**Reply by jarkenau** [reply]
Yes, I forgot to add the `\n`. But after adding them, I run into another error  :confused: 
```
RangeError: Offset is outside the bounds of the DataView
    at DataView.getUint32 (<anonymous>)
    at MessageReader_StandardTypeReader.uint32 (file:///opt/Foxglove%20Studio/resources/app.asar/renderer/3639.js:130:2353)
    at new Record.std_msgs_Header (eval at MessageReader_createParsers (file:///opt/Foxglove%20Studio/resources/app.asar/renderer/3639.js:142:39), <anonymous>:17:23)
```
Again the output file:
[macp-file.zip](https://github.com/foxglove/community/files/10725805/macp-file.zip)

**Reply by jtbandes** [reply]
Your messages have no data:
```
$ mcap cat output-3.mcap 
1676317831042191294 /image [sensor_msgs/Image] []
1676317832049974546 /image [sensor_msgs/Image] []
1676317833074946271 /image [sensor_msgs/Image] []
```

**Reply by jarkenau** [reply]
Thanks for the help, I managed to figure everything out. Do you think a note should be added to the docs, to pay attention that the `\n` are present?

**Reply by jtbandes** [reply]
The documentation states the files should be concatenated, and you can use `gendeps --cat` to produce the correct format. What additional clarification would be helpful? You can make a PR to update the docs if you feel there is a useful change to be made, but personally they seem clear enough to me.
